File: novel-app/server/src/files/outline.ts
// outline.md 读写，照搬 src-tauri/src/files.rs:227-312。
// 格式：# 大纲 / ## 卷名 / - NNN. 标题；章节号补零到 3 位。

import { outlineFile, readFile, writeFileAtomic } from "./atomic.js";
import type { Volume } from "./types.js";

// 写 outline.md，照搬 files.rs:227-237
export function writeOutline(dir: string, outline: Volume[]): void {
  const lines: string[] = ["# 大纲\n"];
  for (const volume of outline) {
    lines.push(`## ${volume.volume}\n`);
    for (const ch of volume.chapters) {
      lines.push(`- ${pad3(ch.num)}. ${ch.title}`);
    }
    lines.push("");
  }
  writeFileAtomic(outlineFile(dir), lines.join("\n"));
}

// Rust format!("{:03}", num)：补零到 3 位
function pad3(n: number): string {
  return String(n).padStart(3, "0");
}

export function readOutline(dir: string): Volume[] {
  const content = readFile(outlineFile(dir));
  if (content === "") return [];
  return parseOutlineText(content);
}

export function getChapterOutline(
  dir: string,
  chapterNum: number,
): string | undefined {
  const outline = readOutline(dir);
  for (const volume of outline) {
    for (const ch of volume.chapters) {
      if (ch.num === chapterNum) return ch.title;
    }
  }
  return undefined;
}

// 解析大纲文本，照搬 files.rs:259-312。三种行格式：## 卷、- N. 标题、裸 N. 标题。
export function parseOutlineText(text: string): Volume[] {
  const outline: Volume[] = [];
  let currentVolume: Volume | null = null;

  for (const rawLine of text.split("\n")) {
    const line = rawLine.trim();
    if (line.startsWith("## ")) {
      if (currentVolume) outline.push(currentVolume);
      currentVolume = { volume: line.slice(3), chapters: [] };
    } else if (line.startsWith("- ") || line.startsWith("* ")) {
      if (currentVolume) {
        const rest = line.slice(2);
        const dotPos = rest.indexOf(". ");
        if (dotPos >= 0) {
          const num = parseInt(rest.slice(0, dotPos).trim(), 10);
          if (!Number.isNaN(num)) {
            currentVolume.chapters.push({
              num,
              title: rest.slice(dotPos + 2).trim(),
            });
          }
        }
      }
    } else {
      // 裸格式：数字. 标题（行首为 ASCII 数字）
      if (currentVolume) {
        const firstChar = line.charAt(0);
        if (firstChar >= "0" && firstChar <= "9") {
          // 跳过前导数字
          const restAfterDigits = line.replace(/^[0-9]+/, "");
          if (restAfterDigits.startsWith(".")) {
            const title = restAfterDigits.slice(1).trimStart();
            if (title !== "") {
              const numEnd = line.search(/[^0-9]/);
              const numStartIdx = numEnd === -1 ? 0 : numEnd;
              const num = parseInt(line.slice(0, numStartIdx), 10);
              if (!Number.isNaN(num)) {
                currentVolume.chapters.push({ num, title });
              }
            }
          }
        }
      }
    }
  }
  if (currentVolume) outline.push(currentVolume);
  return outline;
}

File: app/src-tauri/src/services/files/outline.rs
use std::path::Path;

use super::{outline_file, read_file_content, write_file_atomic};
use crate::domain::types::{ChapterEntry, Volume};
use crate::error::Result;

pub fn write_outline(dir: &Path, outline: &[Volume]) -> Result<()> {
    let mut lines = vec!["# 大纲\n".to_string()];
    for volume in outline {
        lines.push(format!("## {}\n", volume.volume));
        for ch in &volume.chapters {
            lines.push(format!("- {:03}. {}", ch.num, ch.title));
        }
        lines.push(String::new());
    }
    write_file_atomic(&outline_file(dir), &lines.join("\n"))
}

pub fn read_outline(dir: &Path) -> Result<Vec<Volume>> {
    let content = read_file_content(&outline_file(dir))?;
    if content.is_empty() {
        return Ok(Vec::new());
    }
    parse_outline_text(&content)
}

pub fn get_chapter_outline(dir: &Path, chapter_num: u32) -> Result<Option<String>> {
    let outline = read_outline(dir)?;
    for volume in &outline {
        for ch in &volume.chapters {
            if ch.num == chapter_num {
                return Ok(Some(ch.title.clone()));
            }
        }
    }
    Ok(None)
}

pub fn parse_outline_text(text: &str) -> Result<Vec<Volume>> {
    let mut outline: Vec<Volume> = Vec::new();
    let mut current_volume: Option<Volume> = None;

    for line in text.lines() {
        let line = line.trim();
        if line.starts_with("## ") {
            if let Some(vol) = current_volume.take() {
                outline.push(vol);
            }
            current_volume = Some(Volume {
                volume: line.strip_prefix("## ").unwrap_or(line).to_string(),
                chapters: Vec::new(),
            });
        } else if line.starts_with("- ") || line.starts_with("* ") {
            if let Some(ref mut vol) = current_volume {
                let rest = &line[2..];
                if let Some(dot_pos) = rest.find(". ") {
                    if let Ok(num) = rest[..dot_pos].trim().parse::<u32>() {
                        vol.chapters.push(ChapterEntry {
                            num,
                            title: rest[dot_pos + 2..].trim().to_string(),
                        });
                    }
                }
            }
        } else {
            // Format: number. title
            if let Some(ref mut vol) = current_volume {
                if let Some(rest) = line.strip_prefix(|c: char| c.is_ascii_digit()) {
                    let rest = rest.trim_start_matches(|c: char| c.is_ascii_digit());
                    if let Some(rest) = rest.strip_prefix('.') {
                        let rest = rest.trim_start();
                        if !rest.is_empty() {
                            let num_start = line.find(|c: char| !c.is_ascii_digit()).unwrap_or(0);
                            if let Ok(num) = line[..num_start].parse::<u32>() {
                                vol.chapters.push(ChapterEntry {
                                    num,
                                    title: rest.to_string(),
                                });
                            }
                        }
                    }
                }
            }
        }
    }
    if let Some(vol) = current_volume.take() {
        outline.push(vol);
    }
    Ok(outline)
}
